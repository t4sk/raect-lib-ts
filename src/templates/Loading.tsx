import React from "react";
import { Message, Button } from "semantic-ui-react";

interface ErrorProps {
  error: string;
  onClickRetry: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => any;
}

interface Props extends ErrorProps {
  loading: boolean;
  children: () => React.ReactNode;
  renderLoading?: () => React.ReactNode;
  renderError?: (props: ErrorProps) => React.ReactNode;
}

function renderDefaultLoader() {
  return <div>Loading...</div>;
}

const renderDefaultError: React.FC<ErrorProps> = ({ error, onClickRetry }) => {
  return (
    <Message negative>
      <p>{error}</p>
      <Button onClick={(e) => onClickRetry(e)}>Retry</Button>
    </Message>
  );
};

export const Loading: React.FC<Props> = ({
  loading,
  error,
  onClickRetry,
  renderLoading = renderDefaultLoader,
  renderError = renderDefaultError,
  children,
}) => {
  // NOTE: wrap in React fragment to remove ts errors
  if (loading) {
    return <>{renderLoading()}</>;
  }

  if (error) {
    return <>{renderError({ error, onClickRetry })}</>;
  }

  return <>{children()}</>;
};

export default Loading;
