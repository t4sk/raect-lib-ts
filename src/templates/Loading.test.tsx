import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Loading } from "./Loading";

const props = {
  loading: false,
  error: "",
  onClickRetry: jest.fn(),
  children: () => null,
};

beforeEach(() => {
  props.onClickRetry.mockClear();
});

test("it renders", () => {
  const { container } = render(<Loading {...props} />);

  expect(container).toMatchSnapshot();
});

test("loading", () => {
  const { container } = render(<Loading {...props} loading={true} />);

  expect(container).toMatchSnapshot();
});

test("render loading", () => {
  const renderLoading = () => <div>render loading</div>;
  const { container } = render(
    <Loading {...props} loading={true} renderLoading={renderLoading} />
  );

  expect(container).toMatchSnapshot();
});

test("error", () => {
  const { container } = render(<Loading {...props} error="error" />);

  expect(container).toMatchSnapshot();
});

test("render error", () => {
  const renderError = ({ error }: { error: string }) => (
    <div>render {error}</div>
  );
  const { container } = render(
    <Loading {...props} error="error" renderError={renderError} />
  );

  expect(container).toMatchSnapshot();
});

test("onClickRetry", () => {
  const { getByText } = render(<Loading {...props} error="error" />);

  fireEvent.click(getByText("Retry"));

  expect(props.onClickRetry.mock.calls.length).toEqual(1);
});
