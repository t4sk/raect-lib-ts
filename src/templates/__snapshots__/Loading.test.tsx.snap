// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error 1`] = `
<div>
  <div
    class="ui negative message"
  >
    <p>
      error
    </p>
    <button
      class="ui button"
    >
      Retry
    </button>
  </div>
</div>
`;

exports[`it renders 1`] = `<div />`;

exports[`loading 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`render error 1`] = `
<div>
  <div>
    render 
    error
  </div>
</div>
`;

exports[`render loading 1`] = `
<div>
  <div>
    render loading
  </div>
</div>
`;
